mllib/src/main/scala/org/apache/spark/ml/classification/NaiveBayes.scala

```diff
@@ -396,15 +545,29 @@ object NaiveBayesModel extends MLReadable[NaiveBayesModel] {
     private val className = classOf[NaiveBayesModel].getName
 
     override def load(path: String): NaiveBayesModel = {
+      implicit val format = DefaultFormats
       val metadata = DefaultParamsReader.loadMetadata(path, sc, className)
+      val (major, minor) = VersionUtils.majorMinorVersion(metadata.sparkVersion)
+      val modelTypeJson = metadata.getParamValue("modelType")
+      val modelType = Param.jsonDecode[String](compact(render(modelTypeJson)))
 
       val dataPath = new Path(path, "data").toString
       val data = sparkSession.read.parquet(dataPath)
       val vecConverted = MLUtils.convertVectorColumnsToML(data, "pi")
-      val Row(pi: Vector, theta: Matrix) = MLUtils.convertMatrixColumnsToML(vecConverted, "theta")
-        .select("pi", "theta")
-        .head()
-      val model = new NaiveBayesModel(metadata.uid, pi, theta)
+
+      val model = if (major.toInt < 3 || modelType != NaiveBayes.Gaussian) {
```

Comment:
Do we use the Spark version elsewhere like this? rather than a version of the model in the representation?

Reply:
yes, it is also used in:
[LogisticRegressionModel](https://github.com/apache/spark/blob/5853e8b3301fd7b0bff721d5a47139afb17bfd2b/mllib/src/main/scala/org/apache/spark/ml/classification/LogisticRegression.scala#L1229)
[HashingTF](https://github.com/apache/spark/blob/4664a082c2c7ac989e818958c465c72833d3ccfe/mllib/src/main/scala/org/apache/spark/ml/feature/HashingTF.scala#L147)
[LinearRegressionModel](https://github.com/apache/spark/blob/5853e8b3301fd7b0bff721d5a47139afb17bfd2b/mllib/src/main/scala/org/apache/spark/ml/regression/LinearRegression.scala#L769)
and so on

Reply:
I have test loading old version models, and it works fine
In 2.4.4
```scala
scala> import org.apache.spark.ml.feature._
import org.apache.spark.ml.feature._
scala> import org.apache.spark.ml.regression._
import org.apache.spark.ml.regression._
scala> import org.apache.spark.ml.classification._
import org.apache.spark.ml.classification._
scala> var df = spark.read.format("libsvm").load("/data1/Datasets/a9a/a9a")
19/11/09 15:05:36 WARN LibSVMFileFormat: 'numFeatures' option not specified, determining the number of features by going though the input. If you know the number in advance, please specify it via 'numFeatures' option to avoid the extra scan.
df: org.apache.spark.sql.DataFrame = [label: double, features: vector]

scala> df.persist()
res0: org.apache.spark.sql.Dataset[org.apache.spark.sql.Row] = [label: double, features: vector]

scala> df.count
res1: Long = 32561

scala> (0 until 8).foreach(_ => df = df.union(df))

scala> df.count
res3: Long = 8335616                                                            

scala> 

scala> val nb = new NaiveBayes()
nb: org.apache.spark.ml.classification.NaiveBayes = nb_a87b69dac8f6

scala> val model = nb.fit(df)
[Stage 7:==========================================>           (201 + 13) / 256]19/11/09 15:06:03 WARN BLAS: Failed to load implementation from: com.github.fommil.netlib.NativeSystemBLAS
19/11/09 15:06:03 WARN BLAS: Failed to load implementation from: com.github.fommil.netlib.NativeRefBLAS
model: org.apache.spark.ml.classification.NaiveBayesModel = NaiveBayesModel (uid=nb_a87b69dac8f6) with 2 classes

scala> model.save("/tmp/nbm_2.4.4")
```

In this PR:
```scala

scala> import org.apache.spark.ml.classification._
import org.apache.spark.ml.classification._
scala> val model = NaiveBayesModel.load("/tmp/nbm_2.4.4")
model: org.apache.spark.ml.classification.NaiveBayesModel = NaiveBayesModel (uid=nb_a87b69dac8f6) with 2 classes
scala> model.sigma
res0: org.apache.spark.ml.linalg.Matrix = null

scala> model.getModelType
res1: String = multinomial
```